python/tests/expression/test_literals.py

```diff
@@ -0,0 +1,662 @@
+    assert fixed_lit012 == fixed_lit012
+    assert fixed_lit012 != fixed_lit013
+    assert fixed_lit012 < fixed_lit013
+    assert fixed_lit012 <= fixed_lit013
+    assert fixed_lit013 > fixed_lit012
+    assert fixed_lit013 >= fixed_lit012
+
+
+def test_binary_literal():
+    bin_lit012 = literal(bytearray([0x00, 0x01, 0x02]))
+    bin_lit013 = literal(bytearray([0x00, 0x01, 0x03]))
+    assert bin_lit012 == bin_lit012
+    assert bin_lit012 != bin_lit013
+    assert bin_lit012 < bin_lit013
+    assert bin_lit012 <= bin_lit013
+    assert bin_lit013 > bin_lit012
+    assert bin_lit013 >= bin_lit012
+    # None related
+
+
+def test_raise_on_comparison_to_none():
+    bin_lit012 = literal(bytearray([0x00, 0x01, 0x02]))
+    fixed_lit012 = literal(bytes([0x00, 0x01, 0x02]))
+
+    with pytest.raises(AttributeError):
+        bin_lit012 < None
+
+    with pytest.raises(AttributeError):
+        bin_lit012 <= None
+
+    with pytest.raises(AttributeError):
+        bin_lit012 > None
+
+    with pytest.raises(AttributeError):
+        bin_lit012 >= None
+
+    with pytest.raises(AttributeError):
+        fixed_lit012 < None
+
+    with pytest.raises(AttributeError):
+        fixed_lit012 <= None
+
+    with pytest.raises(AttributeError):
+        fixed_lit012 > None
+
+    with pytest.raises(AttributeError):
+        fixed_lit012 >= None
+
+
+def test_binary_to_fixed():
+    lit = literal(bytearray([0x00, 0x01, 0x02]))
+    fixed_lit = lit.to(FixedType(3))
+    assert fixed_lit is not None
+    assert lit.value == fixed_lit.value
+    assert lit.to(FixedType(4)) == FixedLiteral(value=bytearray([0x00, 0x01, 0x02]))
+
+
+def test_raise_on_binary_to_smaller_fixed():
+    lit = literal(bytearray([0x00, 0x01, 0x02]))
+
+    with pytest.raises(ValueError) as exc_info:
+        lit.to(FixedType(2))
+
+    assert "Cannot cast binary to fixed of smaller size: 2" in str(exc_info.value)
+
+
+def test_fixed_to_binary():
+    lit = literal(bytes([0x00, 0x01, 0x02]))
+    binary_lit = lit.to(BinaryType())
+    assert binary_lit is not None
+    assert lit.value == binary_lit.value
+
+
+def test_above_max():
+    a = AboveMax()
+    # singleton
+    assert a == AboveMax()
+    assert str(a) == "AboveMax"
+    with pytest.raises(ValueError) as e:
+        a.value()
+    assert "AboveMax has no value" in str(e.value)
+    with pytest.raises(TypeError) as e:
+        a.to(IntegerType())
+    assert "Cannot change the type of AboveMax" in str(e.value)
+
+
+def test_below_min():
+    b = BelowMin()
+    # singleton
+    assert b == BelowMin()
+    assert str(b) == "BelowMin"
+    with pytest.raises(ValueError) as e:
+        b.value()
+    assert "BelowMin has no value" in str(e.value)
+    with pytest.raises(TypeError) as e:
+        b.to(IntegerType())
+    assert "Cannot change the type of BelowMin" in str(e.value)
+
+
+def test_invalid_boolean_conversions():
+    assert_invalid_conversions(
+        literal(True),
+        [
```

Comment:
This list of types seems to be reused in multiple places, can we refactor to define it once?